ChannelwiseOperator

jakobsj opened the pull request:
This is work in progress but may perhaps be merged soon. Please see description and questions in #78

paskino commented on Wrappers/Python/ccpi/optimisation/operators/ChannelwiseOperator.py:
```diff
@@ -0,0 +1,184 @@
+        for k in range(self.channels):
+            curx = x.subset(channel=k)
+            self.op.adjoint(curx,cury)
+            output_array[k] = cury.as_array()
```
as above

jakobsj replied:
Fixed

jakobsj commented on Wrappers/Python/ccpi/optimisation/operators/ChannelwiseOperator.py:
```diff
@@ -0,0 +1,184 @@
+        for k in range(self.channels):
+            curx = x.subset(channel=k)
+            self.op.direct(curx,cury)
+            output_array[k] = cury.as_array()
```
Ok, fill sounds promising. Do you mean dimension='channel' rather than dimension='angle'? Also how would one specify that one want cury to be filled into channel no k?

jakobsj replied:
Tried this, it runs, but it feels ALL channels with the cury.as_array(), rather than just the `k`th one as desired. This happens both with dimension='angle' and ='channel'. Is there a way to fill on channel `k`?

paskino replied:
oh yes. then could you check that this works as expected?

```python

output.fill(cury.as_array(), channel=k)
```

jakobsj replied:
YES! That does exactly that! Brilliant, thanks!  What a fantastic method, this fill!  I'll update the code with this.  Any ideas for the Q1 on whether there's a more elegant to create a multichannel geometry from a single-channel one?

jakobsj commented:
Updated and believe addressed comments, please have a look again. 

Now using fill as discussed, and have added the out-argument to allow saving memory allocation. The core code of direct and adjoint is repeated due to handling out=None and out!=None, please check if can be avoided.

Remaining point, can the inelegant duplicated code in __init__ to create multichannel versions of geometries be replaced by something neater?

jakobsj commented:
Currently thinking about whether/how ChannelwiseOperator might work for a BlockOperator...

jakobsj commented:
I think making a ChannelwiseOperator from a BlockOperator is not easily possible: ChannelwiseOperator needs to add the channel dimension on to the single-channel operators domain and range geometries, but BlockOperator's geometry BlockGeometry does not have its own dimensions but is merely as list of geometries. 

It is however probably unnecessary, as the same can be achieved by reversing the order and making the BlockOperator from ChannelwiseOperators. 

For this reason I would suggest ChannelwiseOperator to only be supported for non-block operators. This is not a restriction because as discussed above the same functionality can be obtained by a BlockOperator made of the corresponding ChannelwiseOperators.

Perhaps should throw an error in __init__ if single-operator input op is a BlockOperator.

paskino commented:
Please add your considerations in the docstring and add the type test for the operator.

My understanding is that the envisaged usage of this operator is to extend a projection `Operator` (`AcquisitionModel` see SIRF) to work on energy resolved data. As such you shouldn't pass a `BlockOperator` to it. You could put this in a `BlockOperator` and make it work in complex optimisation problems.

Once updated, if Jenkins is ok, please merge.
